src/testing/cmds/harness/test-runner/main.go

```diff
@@ -0,0 +1,182 @@
+	fileName := flag.String("file", "tests_to_run.csv", "csv file containing the list of test names to run generated by the collector.")
+	gotestsumArgsRaw := flag.String("gotestsum-args", "", "Additional arguments to pass to the gotestsum portion of the test command.")
+	gotestArgsRaw := flag.String("gotest-args", "", "Additional arguments to pass to the 'go test' portion of the test command.")
+	shard := flag.Int("shard", 0, "0 indexed current runner index that we are on.")
+	totalShards := flag.Int("total-shards", 1, "Total number of runners that we are sharding over.")
+	threadPool := flag.Int("threads", 1, "Number of tests to execute concurrently.")
+	flag.Parse()
+	gotestsumArgs, err := shlex.Split(*gotestsumArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestsumArgs", zap.Error(err))
+		os.Exit(1)
+	}
+	gotestArgs, err := shlex.Split(*gotestArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestArgs", zap.Error(err))
+		os.Exit(1)
+	}
+
+	err = run(ctx,
+		*tags,
+		*fileName,
+		gotestsumArgs,
+		gotestArgs,
+		*shard,
+		*totalShards,
+		*threadPool,
+	)
+	if err != nil {
+		log.Error(ctx, "Error running tests", zap.Error(err))
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run(ctx context.Context, tags string, fileName string, gotestsumArgs []string, gotestArgs []string, shard int, totalShards int, threadPool int) error {
+	tests, err := readTests(ctx, fileName)
+	if err != nil {
+		return errors.Wrapf(err, "reading file %v", fileName)
+	}
+	slices.Sort(tests) // sort so we shard them from a pre-determined order
+
+	// loop through by the number of shards so that each gets a roughly equal number on each
+	testsForShard := map[string][]string{}
+	for idx := shard; idx < len(tests); idx += totalShards {
+		val := strings.Split(tests[idx], ",")
+		pkg := val[0]
+		testName := val[1]
+		// index all tests by package as we collect the ones for this shard. This lets
+		// us run all tests in each package on this shard with one `go test` command, preserving the serial
+		// running of tests without t.parallel the same way that go test ./... would since
+		// got test also runs packages in paralllel.
+		if _, ok := testsForShard[pkg]; !ok {
+			testsForShard[pkg] = []string{testName}
+		} else {
+			testsForShard[pkg] = append(testsForShard[pkg], testName)
+		}
+	}
+
+	eg, _ := errgroup.WithContext(ctx)
+	eg.SetLimit(threadPool)
+	for pkg, tests := range testsForShard {
+		threadLocalPkg := pkg
+		threadLocalTests := tests
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		eg.Go(func() error {
+			return errors.EnsureStack(runTest(threadLocalPkg, threadLocalTests, tags, gotestsumArgs, gotestArgs))
+		})
+	}
+	err = eg.Wait()
+
+	if err != nil {
+		return errors.EnsureStack(err)
+	}
+	return nil
+}
+
+func readTests(ctx context.Context, fileName string) ([]string, error) {
+	lockFileName := fmt.Sprintf("lock-%s", fileName)
```

Comment:
Consider fsnofity here: https://github.com/fsnotify/fsnotify

If you adjust the test computer thingie to atomically write the file, then you just need to notify on the file existing.  And, fsnotify will react instantly to the creation of the file, so there is no delay between creating the file and the tests starting.  (backoff.RetryNotify is probably napping at the exact instant.)